### ISSUE TITLE
typed-d3 vs DefinitelyTyped/d3

### ISSUE THREAD

**Issue by myitcv** [typed-typings/npm-d3#1]
At the time of writing, the diff between the type definition in this repo and that on DefinitelyTyped is:

``` diff
--- DefinitelyTyped/d3/d3.d.ts  2015-12-02 10:18:19.698205131 +0000
+++ typings/typed-d3/index.d.ts 2015-12-02 10:02:38.989999203 +0000
@@ -2697,24 +2697,24 @@
         response(): (request: XMLHttpRequest) => any;
         response(value: (request: XMLHttpRequest) => any): DsvXhr<T>;

-        get(callback?: (err: XMLHttpRequest, data: T[]) => void): DsvXhr<T>;
-        post(data?: any, callback?: (err: XMLHttpRequest, data: T[]) => void): DsvXhr<T>;
-        post(callback: (err: XMLHttpRequest, data: T[]) => void): DsvXhr<T>;
+    get(callback?: (err: any, data: T) => void): DsvXhr<T>;
+    post(data?: any, callback?: (err: any, data: T) => void): DsvXhr<T>;
+    post(callback: (err: any, data: T) => void): DsvXhr<T>;

-        send(method: string, data?: any, callback?: (err: XMLHttpRequest, data: T[]) => void): DsvXhr<T>;
-        send(method: string, callback: (err: XMLHttpRequest, data: T[]) => void): DsvXhr<T>;
+    send(method: string, data?: any, callback?: (err: any, data: T) => void): DsvXhr<T>;
+    send(method: string, callback: (err: any, data: T) => void): DsvXhr<T>;

         abort(): DsvXhr<T>;

         on(type: "beforesend"): (request: XMLHttpRequest) => void;
         on(type: "progress"): (request: XMLHttpRequest) => void;
-        on(type: "load"): (response: T[]) => void;
+    on(type: "load"): (response: T) => void;
         on(type: "error"): (err: any) => void;
         on(type: string): (...args: any[]) => void;

         on(type: "beforesend", listener: (request: XMLHttpRequest) => void): DsvXhr<T>;
         on(type: "progress", listener: (request: XMLHttpRequest) => void): DsvXhr<T>;
-        on(type: "load", listener: (response: T[]) => void): DsvXhr<T>;
+    on(type: "load", listener: (response: T) => void): DsvXhr<T>;
         on(type: "error", listener: (err: any) => void): DsvXhr<T>;
         on(type: string, listener: (...args: any[]) => void): DsvXhr<T>;
     }
@@ -3313,6 +3313,4 @@
 // we need this to exist
 interface TouchList { }

-declare module 'd3' {
     export = d3;
-}
```

This highlights the fact that this repo's copy has fallen behind in a couple of places.... and because there are two places to update, that is always a possibility.

But ignoring that for one second, the two are largely identical (and intended as such) with the exception of:

``` diff
-declare module 'd3' {
     export = d3;
-}
```

Why not split the DefinitelyTyped definition and drop the copy in this repo, thereby creating a single place for modifying what is essentially a shared definition?

I've pushed up an example of what I mean: https://github.com/myitcv/DefinitelyTyped/tree/d3_split_example/d3

The typings registry would then reference the [`d3-typings.d.ts`](https://github.com/myitcv/DefinitelyTyped/blob/d3_split_example/d3/d3-typings.d.ts) file (which would transitively pull in [`d3-common.d.ts`](https://github.com/myitcv/DefinitelyTyped/blob/d3_split_example/d3/d3-common.d.ts)). 

Thoughts?

**Comment by blakeembrey** [comment]
@myitcv You're absolutely right. I intended to publish support for https://github.com/typings/typings/issues/51 to solve this, actually. That way, DefinitelyTyped can remain ambient and we just run `typings bundle` for them.

However, references aren't currently resolved inline using `typings` (to hard to figure out intent, so I'm planning on leaving that to the user), so that separation doesn't work just yet. You should be able to export from the main definition though.

**Comment by myitcv** [comment]
> You should be able to export from the main definition though.

Can you elaborate on what you mean here?

**Comment by blakeembrey** [comment]
**d3-common.d.ts** (rename to `d3-typings.d.ts`)

```
...

export = d3;
```

**d3.d.ts**

```
declare module "d3" {
  import main = require('./d3-typings');
  export = main;
}
```

This should work. In **d3.d.ts** you might also need a reference, but I've used this trick in TypeScript 1.5 to support NPM modules I published which having to re-compile the `.d.ts` files generated by TypeScript.

**Comment by blakeembrey** [comment]
@myitcv Picking back up from earlier, DefinitelyTyped ambient modules can now be generated from an external module declaration using `typings bundle`. That might help, since it's easy to go that way but difficult in reverse.

**Comment by twastvedt** [comment]
@blakeembrey Do I gather from your last comment that ideally the d3 typings would only live here? Sorry, I don't quite understand the nuances of type definitions yet. I'd like to add typings for a couple recent features, for example `zoom.xExtent` and `zoom.yExtent`. Should I submit a pull request just to this repo, or to here and DefinitelyTyped? Thanks!

**Comment by twastvedt** [comment]
Well, I feel a little silly. The referenced changes (zoom.xExtent) were made to a branch that is not yet in the master. I'm still interested in the answer to my question above, though, as the differences between this and DefinitelyTyped have increased since the start of this issue.

**Comment by blakeembrey** [comment]
@twastvedt It's really up to you 😄 In Typings, maintenance is easier. Would you like to own this repo? It doesn't seem the current maintainer is making changes.

**Comment by twastvedt** [comment]
I'd like to contribute, but I'm not sure maintaining would be the best role for me. I don't use d3 heavily, and only sporadically. 

To push the question further, I'm still not clear on the best approach here. Your comment:

> DefinitelyTyped ambient modules can now be generated from an external module declaration using typings bundle. That might help, since it's easy to go that way but difficult in reverse.

seems to suggest that it would be most straightforward to just keep this repo up-to-date, and generate ambient typings using this repo and `typings bundle`. If I make improvements to the DT typings, however, those would then need to be applied here as well, right? So, wouldn't it be ideal for typings to live here only, instead of needing to update two repos with every change? Or am I just missing something?
